segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/exts3/ExtS3Storage.java

```diff
@@ -0,0 +1,445 @@
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return supplyAsync(streamSegmentName, () -> syncOpenWrite(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncCreate(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
+            timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncWrite(handle, offset, data, length));
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncSeal(handle));
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+        return supplyAsync(targetHandle.getSegmentName(),
+                () -> syncConcat(targetHandle, offset, sourceSegment));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncDelete(handle));
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+
+    }
+
+    //endregion
+
+    //region private sync implementation
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openRead", streamSegmentName);
+
+        StreamSegmentInformation info = syncGetStreamSegmentInfo(streamSegmentName);
+        ExtS3SegmentHandle retHandle = ExtS3SegmentHandle.getReadHandle(streamSegmentName);
+        LoggerHelpers.traceLeave(log, "openRead", traceId, streamSegmentName);
+        return retHandle;
+    }
+
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        StreamSegmentInformation info = syncGetStreamSegmentInfo(streamSegmentName);
+        ExtS3SegmentHandle retHandle;
+        if (info.isSealed()) {
+            retHandle = ExtS3SegmentHandle.getReadHandle(streamSegmentName);
+        } else {
+            retHandle = ExtS3SegmentHandle.getWriteHandle(streamSegmentName);
+        }
+
+        LoggerHelpers.traceLeave(log, "openWrite", traceId);
+        return retHandle;
+    }
+
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) throws IOException {
+        long traceId = LoggerHelpers.traceEnter(log, "read", handle.getSegmentName(), offset, bufferOffset, length);
+
+        if (offset < 0 || bufferOffset < 0 || length < 0) {
+            throw new CompletionException(new ArrayIndexOutOfBoundsException());
```

Comment:
Please either use `@sneakyThrows` or have the method declare an explicit exception and only wrap this exception at the very top. The same applies for all other exceptions in this class.